dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/relationship/DefaultRelationshipService.java

```diff
@@ -0,0 +1,445 @@
+        result.setTrackedEntityType( trackedEntityType );
+        result.setCreated( DateUtils.parseDate( tei.getCreated() ) );
+        result.setCreatedAtClient( DateUtils.parseDate( tei.getCreatedAtClient() ) );
+        result.setLastUpdated( DateUtils.parseDate( tei.getLastUpdated() ) );
+        result.setLastUpdatedAtClient( DateUtils.parseDate( tei.getLastUpdatedAtClient() ) );
+        result.setInactive( tei.isInactive() );
+        result.setGeometry( tei.getGeometry() );
+        result.setDeleted( tei.isDeleted() );
+        result.setPotentialDuplicate( tei.isPotentialDuplicate() );
+        result.setStoredBy( tei.getStoredBy() );
+        result.setCreatedByUserInfo( tei.getCreatedByUserInfo() );
+        result.setLastUpdatedByUserInfo( tei.getLastUpdatedByUserInfo() );
+
+        // NOTE: skipping mapping of trackedEntity.relationships as /tracker/relationships models do not include the
+        // relationships fields of each of its nested entities as relationship items would then end in an infinite recursion
+
+        result.setProgramInstances( tei.getEnrollments().stream().map( this::map ).collect( Collectors.toSet() ) );
+        result.setProgramOwners( tei.getProgramOwners().stream().map( this::map ).collect( Collectors.toSet() ) );
+        result.setTrackedEntityAttributeValues(
+            tei.getAttributes().stream().map( this::map ).collect( Collectors.toSet() ) );
+
+        return result;
+    }
+
+    private TrackedEntityProgramOwner map( org.hisp.dhis.dxf2.events.trackedentity.ProgramOwner programOwner )
+    {
+        TrackedEntityInstance trackedEntityInstance = new TrackedEntityInstance();
+        trackedEntityInstance.setUid( programOwner.getTrackedEntityInstance() );
+
+        Program program = new Program();
+        program.setUid( programOwner.getProgram() );
+
+        OrganisationUnit organisationUnit = new OrganisationUnit();
+        organisationUnit.setUid( programOwner.getOwnerOrgUnit() );
+
+        return new TrackedEntityProgramOwner( trackedEntityInstance, program, organisationUnit );
+    }
+
+    private TrackedEntityAttributeValue map( org.hisp.dhis.dxf2.events.trackedentity.Attribute att )
+    {
+        TrackedEntityAttribute attribute = new TrackedEntityAttribute();
+        attribute.setUid( att.getAttribute() );
+        attribute.setCreated( DateUtils.parseDate( att.getCreated() ) );
+        attribute.setLastUpdated( DateUtils.parseDate( att.getLastUpdated() ) );
+        attribute.setValueType( att.getValueType() );
+        attribute.setCode( att.getCode() );
+        attribute.setSkipSynchronization( att.isSkipSynchronization() );
+        TrackedEntityAttributeValue result = new TrackedEntityAttributeValue();
+        result.setAttribute( attribute );
+        result.setValue( att.getValue() );
+        result.setStoredBy( att.getStoredBy() );
+        return result;
+    }
+
+    private ProgramInstance map( org.hisp.dhis.dxf2.events.enrollment.Enrollment enrollment )
```

Comment:
This dependency to dxf2, along with the two in the methods above, are only here because we still have no service that maps from the view to dhis-api directly, but when we do, we'll remove this, right?

Reply:
exactly 😄